cmd/geoipupdate/main.go

```diff
@@ -0,0 +1,428 @@
+	if err != nil {
+		return errors.Wrap(err, "error creating request")
+	}
+	if !strings.HasPrefix(editionID, "GeoLite2") {
+		req.SetBasicAuth(fmt.Sprintf("%d", config.AccountID), config.LicenseKey)
+	}
+
+	if verbose {
+		log.Printf("Performing update request to %s", url)
+	}
+	res, err := client.Do(req)
+	if err != nil {
+		return errors.Wrap(err, "error performing HTTP request")
+	}
+	defer func() {
+		if err := res.Body.Close(); err != nil {
+			log.Fatalf("Error closing response body: %+v", errors.Wrap(err, "closing body"))
+		}
+	}()
+
+	if res.StatusCode == http.StatusNotModified {
+		if verbose {
+			log.Printf("No new updates available for %s", editionID)
+		}
+		return nil
+	}
+
+	if res.StatusCode != http.StatusOK {
+		buf, err := ioutil.ReadAll(res.Body)
+		if err == nil {
+			return errors.Errorf("unexpected HTTP status code: %s: %s", res.Status, buf)
+		}
+		return errors.Errorf("unexpected HTTP status code: %s", res.Status)
+	}
+
+	newMD5 := res.Header.Get("X-Database-MD5")
+	if newMD5 == "" {
+		return errors.New("no X-Database-MD5 header found")
+	}
+	lastModified, err := getLastModified(res.Header)
+	if err != nil {
+		return err
+	}
+
+	return writeAndCheck(config, verbose, filename, res.Body, newMD5, lastModified)
+}
+
+func getLastModified(
+	headers http.Header,
+) (time.Time, error) {
+	lastModifiedStr := headers.Get("Last-Modified")
+	if lastModifiedStr == "" {
+		return time.Time{}, errors.New("no Last-Modified header found")
+	}
+
+	t, err := time.ParseInLocation(time.RFC1123, lastModifiedStr, time.UTC)
+	if err != nil {
+		return time.Time{}, errors.Wrap(err, "error parsing time")
+	}
+
+	return t, nil
+}
+
+func writeAndCheck(
+	config *Config,
+	verbose bool,
+	filename string,
+	body io.Reader,
+	newMD5 string,
+	lastModified time.Time,
+) error {
+	targetTest := filepath.Join(
+		config.DatabaseDirectory,
+		fmt.Sprintf("%s.test", filename),
+	)
+
+	fh, err := os.OpenFile(targetTest, os.O_WRONLY|os.O_CREATE|os.O_TRUNC, 0644)
+	if err != nil {
+		return errors.Wrap(err, "error creating file")
+	}
+
+	gzReader, err := gzip.NewReader(body)
+	if err != nil {
+		_ = fh.Close()
+		_ = os.Remove(targetTest)
+		return errors.Wrap(err, "error creating gzip reader")
+	}
+
+	md5Writer := md5.New()
+	multiWriter := io.MultiWriter(fh, md5Writer)
+
+	if _, err := io.Copy(multiWriter, gzReader); err != nil {
+		_ = fh.Close()
+		_ = os.Remove(targetTest)
+		_ = gzReader.Close()
+		return errors.Wrap(err, "error reading/writing")
+	}
+
+	if err := gzReader.Close(); err != nil {
+		_ = fh.Close()
+		_ = os.Remove(targetTest)
+		return errors.Wrap(err, "error closing gzip reader")
+	}
+
+	if err := fh.Sync(); err != nil {
+		_ = fh.Close()
+		_ = os.Remove(targetTest)
+		return errors.Wrap(err, "error syncing file")
+	}
+
+	if err := fh.Close(); err != nil {
+		_ = os.Remove(targetTest)
+		return errors.Wrap(err, "error closing file")
+	}
+
+	gotMD5 := fmt.Sprintf("%x", md5Writer.Sum(nil))
+	if gotMD5 != newMD5 {
```

Comment:
Maybe `strings.EqualFold` in case the case of one of them changes for some reason?

Reply:
May as well!